src/Adapter/MSTest.CoreAdapter/Execution/TestExecutionManager.cs

```diff
@@ -220,96 +331,55 @@ private void ExecuteTestsInSource(IEnumerable<TestCase> tests, IRunContext runCo
-                    // Skip test if not fitting filter criteria.
-                    if (filterExpression != null && filterExpression.MatchTestCase(
-                            currentTest,
-                            (p) => this.TestMethodFilter.PropertyValueProvider(currentTest, p)) == false)
-                    {
-                        continue;
-                    }
-
-                    // Send previous test result.
-                    this.SendTestResults(test, unitTestResult, startTime, endTime, testExecutionRecorder);
-
-                    if (this.cancellationToken != null && this.cancellationToken.Canceled)
-                    {
-                        break;
-                    }
-
-                    var unitTestElement = currentTest.ToUnitTestElement(source);
-                    testExecutionRecorder.RecordStart(currentTest);
+                    break;
+                }
 
-                    startTime = DateTimeOffset.Now;
+                var unitTestElement = currentTest.ToUnitTestElement(source);
+                testExecutionRecorder.RecordStart(currentTest);
 
-                    PlatformServiceProvider.Instance.AdapterTraceLogger.LogInfo(
-                        "Executing test {0}",
-                        unitTestElement.TestMethod.Name);
+                var startTime = DateTimeOffset.Now;
 
-                    // this is done so that appropriate values of testcontext properties are set at source level
-                    // and are merged with session level parameters
-                    var sourceLevelParameters = PlatformServiceProvider.Instance.SettingsProvider.GetProperties(source);
+                PlatformServiceProvider.Instance.AdapterTraceLogger.LogInfo(
+                    "Executing test {0}",
+                    unitTestElement.TestMethod.Name);
 
-                    if (this.sessionParameters != null && this.sessionParameters.Count > 0)
-                    {
-                        sourceLevelParameters = sourceLevelParameters.Concat(this.sessionParameters).ToDictionary(x => x.Key, x => x.Value);
-                    }
+                var unitTestResult = testRunner.RunSingleTest(unitTestElement.TestMethod, sourceLevelParameters);
 
-                    unitTestResult = testRunner.RunSingleTest(unitTestElement.TestMethod, sourceLevelParameters);
+                PlatformServiceProvider.Instance.AdapterTraceLogger.LogInfo(
+                    "Executed test {0}",
+                    unitTestElement.TestMethod.Name);
 
-                    PlatformServiceProvider.Instance.AdapterTraceLogger.LogInfo(
-                        "Executed test {0}",
-                        unitTestElement.TestMethod.Name);
+                var endTime = DateTimeOffset.Now;
 
-                    endTime = DateTimeOffset.Now;
-                    test = currentTest;
-                }
+                this.SendTestResults(currentTest, unitTestResult, startTime, endTime, testExecutionRecorder);
```

Comment:
Is `SendTestResults` thread safe?

Reply:
It mostly boils down to the RecordEnd and RecordResult APIs. Everything else is stateless. I believe those APIs in TP are thread safe as well.